File: Data Structures & Algorithms/depth-of-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {

        const _dfs = (root, depth=1) => {
            if (!root) return depth - 1;

            return Math.max(
                _dfs(root.left, depth+1),
                _dfs(root.right, depth+1)
            );
        }
        return _dfs(root, 1);
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        def get_max_height(node):
            if not node:
                return 0
            return 1 + max(get_max_height(node.left), get_max_height(node.right))
        
        left = get_max_height(root.left)
        right = get_max_height(root.right)

        if abs(left - right) > 1:
            return False
        
        return self.isBalanced(root.left) and self.isBalanced(root.right)
            



